docs/blog/2023-06-10-encoder-refactors.md

```diff
@@ -0,0 +1,119 @@
+
+## Configuration Changes
+
+The major changes to configuration in the devicetree files relates to how the number of steps/triggers for a given encoder are set. In particular, the number of pulses/steps for a given encoder is configured first, allowing ZMK to determine the exact angular degrees of change that is represented by a single pulse on the data lines to that encoder.
+
+Once that angular degrees mapping is completed, now independently there is a configuration setting to control how many triggers of the behavior in the keymap should occur for each full rotation of the sensor. Another way to think of this is "how many degrees of rotation results in a triggering of the sensor behavior in your keymap layer.
+
+Splitting these two parts of the encoder configuration allows greater flexibility, and fine grained control of encoder behavior for linear encoders that don't have fixed detents.
+
+### Old Configuration
+
+Previously, an encoder configuration looked like:
+
+```
+    left_encoder: encoder_left {
+        compatible = "alps,ec11";
+        label = "LEFT_ENCODER";
+        a-gpios = <&pro_micro 21 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>;
+        b-gpios = <&pro_micro 20 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>;
+        resolution = <4>;
+    };
+```
+
+Here, the `resolution` property was used to indicate how many encoder pulses should trigger the sensor behavior one time. Next, the encoder is selected in the sensors node:
+
+```
+    sensors {
+        compatible = "zmk,keymap-sensors";
+        sensors = <&left_encoder &right_encoder>;
+    };
+```
+
+That was the entirety of the configuration for encoders.
+
+### New Configuration
+
+```
+    left_encoder: encoder_left {
+        compatible = "alps,ec11";
+        label = "LEFT_ENCODER";
+        a-gpios = <&pro_micro 21 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>;
+        b-gpios = <&pro_micro 20 (GPIO_ACTIVE_HIGH | GPIO_PULL_UP)>;
+        steps = <20>;
+    };
+```
+
+Here, the `steps` property is now used to indicate how many encoder pulses there are in a single complete rotation of the encoder. Next, the encoder is selected in the sensors node as before, but an additional configuration is used to indicate how many times the encoder should trigger the behavior in your keymap per rotation:
+
+```
+    sensors {
+        compatible = "zmk,keymap-sensors";
+        sensors = <&left_encoder &right_encoder>;
+        triggers-per-rotation = <20>;
+    };
+```
+
+For tactile encoders that have detents, the `triggers-per-rotation` would match the number of detents on the encoder. For linear encoders, the value can be chosen to suit your needs.
+
+## Zephyr Sensor Drivers
+
+The configuration changes bring ZMK's code in line with how upstream Zephyr sensor drivers handle rotations. This has the added advantage of allowing us to leverage other sensor drivers. On Nordic MCUs, like nRF52840, the NRFX QDEC driver can be used, for example:
```

Comment:
I guess this is not strictly a docs question, but it might be good to clarify here: Can you define a "generic" encoder node for a shield like the current implementation, without being nrf-specific? Since we now support RP2040 controllers in pro micro form factor, it might be an issue and a regression compared to the status quo.

If you can still define a generic version, maybe that can be overridden in `boards/nice_nano.overlay` etc. for nRF boards.

Reply:
The *current* `alps,ec11` (which is admittedly a bad compatible name, since quadrature encoders are not specific to alps *at all*) driver is MCU agnostic, and preferable. If you want to use the NRFX QDEC driver for a generic shield, you most definitely would use the `alps,ec11` driver in the base `<shield>.overlay`, and then override in a board specific file.

You can see an example of this in the ZMK Uno shield changes in this PR.

Reply:
Thanks, that makes sense. I misread the blog post and conflated the update for the generic driver with the *suggestion* of utilizing the NRFX QDEC driver.